### PR TITLE
Add `like` and `str` to matcher

### PR DESCRIPTION
**Pull request by smcclellan**
This is a refactored version of https://github.com/puppetlabs/razor-server/pull/186, which contains the original commits and feedback. I tweaked the original commits some to add testing and simplify the implementation.

This adds two functions, `like` and `str`, for performing regular expression matching.

Fixes https://tickets.puppetlabs.com/browse/RAZOR-500

### CODE REVIEW

**Comment by lutter** [comment]
ACK. Looks good

File: spec/matcher_spec.rb [hunk 1]
```diff
@@ -132,6 +132,21 @@ def match_metadata(*rule)
       match("=", "[]", "[]").should == true
     end
 
+    describe "like function" do
+      [['abc', 'abc'], ['abc', 'a.c'], ['abc', 'a.+'], ['abc', 'a.*'],
+       ['abc', 'ab'], ['abc', ''], ['abc', '\Aabc\z'], ['abc', '^abc$']].each do |str, reg|
```
Comment (smcclellan) [inline spec/matcher_spec.rb]: Add cases for invalid regular expressions.

File: spec/matcher_spec.rb [hunk 2]
```diff
@@ -228,6 +265,22 @@ def match_metadata(*rule)
...
+      Matcher.new(["like", "abc", "def"]).should be_valid
+    end
+
+    it "should allow strings for 'like' function" do
+      Matcher.new(["like", true, false]).should_not be_valid
+      Matcher.new(["like", 1, 2]).should_not be_valid
+      Matcher.new(["like", "abc", false]).should_not be_valid
+      Matcher.new(["like", 1, "abc"]).should_not be_valid
+      Matcher.new(["like", "abc", "def"]).should be_valid
+    end
```
Comment (smcclellan) [inline spec/matcher_spec.rb]: Duplicate code.

File: spec/matcher_spec.rb (hunk 1 above)
Comment (smcclellan) [inline spec/matcher_spec.rb]: (Resolved.)

File: spec/matcher_spec.rb (hunk 2 above)
Comment (smcclellan) [inline spec/matcher_spec.rb]: (Resolved.)

File: lib/razor/matcher.rb [hunk 3]
```diff
@@ -306,5 +318,13 @@ def validate(rule, required_returns, caller_name = nil, caller_position = nil)
         end
       end
     end
+
+    if name == 'like' and errors.empty?
+      begin
+        Regexp.new(rule[2])
+      rescue RegexpError => e
+        errors << _("invalid regular expression supplied to `like` for argument %{position}") % {position: 1}
```
Comment (samwoods1) [inline lib/razor/matcher.rb]: You might want to include e.message in the text as well?  It might give more context to a user about the specific error.

Comment (smcclellan) [inline lib/razor/matcher.rb]: Good idea, that could help.

File: spec/matcher_spec.rb [hunk 4]
```diff
@@ -168,6 +190,28 @@ def match_metadata(*rule)
...
+        match("=", ["str", "2.7"], "2.7").should == true
+        match("=", ["str", "1e5"], "1e5").should == true
+      end
+
+      it "should behave for valid booleans and nil" do
+        match("=", ["str", true ], "true").should == true
+        match("=", ["str", false], "false").should == true
+        match("=", ["str", nil], '').should == true
+      end
+    end
```
Comment (aperiodic) [inline spec/matcher_spec.rb]: Is there anything `str` is not expected to work on?

Comment (smcclellan) [inline spec/matcher_spec.rb]: Not at this time. We have no complex objects or collections that can be used as arguments.

Perhaps worth noting, structured facts/metadata will throw a wrench in the matcher, but that's an [existing issue](https://tickets.puppetlabs.com/browse/RAZOR-641).

**Comment by smcclellan** [comment]
Rebased and merged.

**Pull request merged**